pkg/wwan/mmagent/agent.go

```diff
@@ -0,0 +1,1451 @@
+	//       brings in tons of unnecessary dependencies. It would be better to move this
+	//       constant to pillar/types or to some other small common package.
+	//       Alternatively, CheckAndGetNetworkProxy (which has those many deps) could be
+	//       moved out from devicenetwork to some other place.
+	WwanResolvConfDir = "/run/wwan/resolv.conf"
+)
+
+var (
+	_, ipv4Any, _    = net.ParseCIDR("0.0.0.0/0")
+	_, ipv6Any, _    = net.ParseCIDR("::/0")
+	emptyPhysAddrs   = types.WwanPhysAddrs{}
+	defaultProbeAddr = net.ParseIP("8.8.8.8")
+	emptyIPSettings  = types.WwanIPSettings{}
+)
+
+// Version is set from Makefile
+var Version = "No version specified"
+
+// MMAgent is an EVE microservice controlling ModemManager (https://modemmanager.org/).
+type MMAgent struct {
+	agentbase.AgentBase
+	logger     *logrus.Logger
+	log        *base.LogObject
+	ps         *pubsub.PubSub
+	versionPtr *bool
+
+	// publications
+	pubWwanStatus        pubsub.Publication
+	pubWwanMetrics       pubsub.Publication
+	pubWwanLocationInfo  pubsub.Publication
+	pubCipherBlockStatus pubsub.Publication
+	pubCipherMetrics     pubsub.Publication
+	cipherMetrics        *cipher.AgentMetrics
+
+	// subscriptions
+	subGlobalConfig   pubsub.Subscription
+	subWwanConfig     pubsub.Subscription
+	subControllerCert pubsub.Subscription
+	subEdgeNodeCert   pubsub.Subscription
+
+	// client for communication with MM
+	mmClient *mmdbus.Client
+
+	// global config properties
+	gcInitialized     bool
+	globalConfig      types.ConfigItemValueMap
+	dpcKey            string
+	dpcTimestamp      time.Time
+	rsConfigTimestamp time.Time
+	radioSilence      bool
+	locPublishPeriod  time.Duration
+	locTrackingModem  string // selected modem for location tracking (DBus path)
+	scanProviders     bool
+
+	// config, state data and metrics collected for every cellular modem
+	modemInfo     map[string]*ModemInfo // key: DBus path
+	missingModems []types.WwanNetworkConfig
+
+	// True when modem metrics have been updated and should be published
+	metricsUpdated bool
+}
+
+// ModemInfo : collection of config, state data and metrics stored by the agent
+// inside MMAgent.modemInfo for every modem detected by ModemManager.
+// Note that modems which have config but are not physically present are recorded
+// in MMAgent.missingModems.
+type ModemInfo struct {
+	// State data and metrics received from the ModemManager D-Bus client.
+	mmdbus.Modem
+	// Unmanaged modem has empty Config (LogicalLabel is empty string).
+	config types.WwanNetworkConfig
+	// Previous config - used only within the applyWwanConfig function.
+	prevConfig types.WwanNetworkConfig
+	// IP settings applied for the wwan* interface in the Linux network stack.
+	appliedIPSettings types.WwanIPSettings
+	// Decrypted username and password (from Config.AccessPoint.EncryptedCredentials).
+	decryptedUsername string
+	decryptedPassword string
+	// Latest errors encountered while managing this modem.
```

Comment:
Is it useful during debugging to see just last error ? Although I see it's separated by actions (probe, connect, decrypt, locTracking), but why not store slice of errors?

Reply:
Previous errors will be found in logs, but for publishing the latest WwanStatus we only need the last error (per action).

Reply:
Got it, thanks!